### DISCUSSION TITLE
Mitigate "Augmix" from mmclassification to mmpretrain

### DISCUSSION THREAD

**Discussion by EdyQiu** [open-mmlab/mmclassification#1511] [Q&A]
Hello, I encountered an issue while trying to migrate the "Augmix" class from the Misclassification to the MMpretrain. "Augmix" is an augmentation method that I wrote myself and placed in the PIPELINE module. 
![image](https://user-images.githubusercontent.com/23387293/233947891-0b1a1610-e414-42fc-aa1d-4402c23fb7e4.png)

The implementation is straightforward, but when loading 1 image, 3 augmented images will be the output. it returns a list instead of a dictionary, which causes issues in the data-loading pipeline.
![image](https://user-images.githubusercontent.com/23387293/233948184-ecdf386b-54fe-4800-a4cb-32082aa5a2da.png)

To resolve this issue, I modified the "Compose" class in the data-loading pipeline to accept a list as input. 
![image](https://user-images.githubusercontent.com/23387293/233948823-4e34c9bd-475b-42ac-b75e-4d457a2dfeaf.png)

However, when attempting to migrate the code from Misclassification to MMpretrain, I encountered several problems.

The most significant issue is that I could not find an equivalent function to "Compose" in mmpretrain. As a result, I am unable to resolve the problem of "Augmix" returning a list instead of a dictionary, causing 'PackInputs' to fail to work.

So I am here to ask for some help, is there an equivalent function to "Compose" in mmpretrain. Or is there any clever idea to implement "Augmix" in MMpretrain. 

Thanks a lot for the help!

**Comment by EdyQiu** [comment]
Sorry I used the wrong word "Mitigate", it should be "Migrate"

**Comment by EdyQiu** [comment]
Another problem: image normalization is conducted in Augmix, soit has conflict with "data_preprocessor", but I guess I can simply not use the data_preprocessor, and use "normalize" in data pipeline if needed.

**Comment by Ezra-Yu** [comment]
I'm very sorry for the inconvenience caused by this upgrade. Compse is in mmcv. Specifically, you can access https://github.com/open-mmlab/mmcv/blob/89a264527e3dc9c5eebed6195faa709d446c7a9c/mmcv/transforms/wrappers.py#L40 

to disuse normalize in data_preprocessor:
```
data_preprocessor = dict(
    # RGB format normalization parameters
    mean=None,
    std=None,
    # convert image from BGR to RGB
    to_rgb=True,
)
```


I notice that Augmix has been supported in torchvision refer to: https://pytorch.org/vision/stable/generated/torchvision.transforms.AugMix.html#torchvision.transforms.AugMix

Maybe you can use that, refer to PR #1265

**Reply by EdyQiu** [reply]
Thank you for your previous advice! I have decided to implement Augmix myself to facilitate further modifications. However, I am still struggling with the issue of multiple image outputs during the training pipeline. When loading one image, the actual batch has three additional images.

I attempted to set a breakpoint in "Compose" in mmcv, but it did not stop there. I am now unsure what to try next.

I apologize for bothering you again, but could you please offer any suggestions or guidance?

**Reply by Ezra-Yu** [reply]
why not just return a mixed image in `Augmix` and put some params in result['augmix_param'].

I can not get why you want to return the list of images.

**Reply by EdyQiu** [reply]
While since Augmix  is a module in the loading PIPELINE, it should return **result**, which is a dictionary with keys(['img_path', 'gt_label', 'sample_idx', 'img', 'img_shape', 'ori_shape', 'flip', 'flip_direction']), which contains **one image only**
![image](https://user-images.githubusercontent.com/23387293/233981792-695016e4-2b0a-45cc-8cb4-b822f8415d01.png)
but Augmix should return **3 images**: the original copy, the first augmentation, the second augmentation(they are used to compute consistency loss). So I am confused about how to return 3 images.
In mmclassification, I return a list with 3 "results" to return all 3 images, and modified "Compose", so that 3 elements in "return" will come to the next module (e.g. "ImageToTensor") **one by one**
![image](https://user-images.githubusercontent.com/23387293/233982293-0ba4d89f-f5a7-48fa-ba09-651ef24065f3.png)

Would you mind explaining further about "return a mixed image in Augmix and put some params in result['augmix_param']." ?

**Comment by EdyQiu** [comment]
I solved the problem by modifying the following files in mmpretrain, to replace the equivalent parts in "Compose" in mmclassification:
![image](https://user-images.githubusercontent.com/23387293/233995043-b86b8851-9d9d-4076-8dec-f73c8f8a681c.png)
![image](https://user-images.githubusercontent.com/23387293/233995103-e11f902f-70eb-44d4-8cbc-aeea94a6476a.png)




Thanks for your help !
